pycbc/inference/models/tools.py

```diff
@@ -193,82 +247,293 @@ def marginalize_loglr(self, sh_total, hh_total,
+            logweight += snrv.squared_norm().numpy()
+        logweight /= 2.0
+
+        # Draw proportional to the incoherent likelihood
+        # Draw first which time sample
+        tci = draw_sample(logweight, size=self.vsamples)
+        # Second draw a subsample size offset so that all times are covered
+        tct = numpy.random.uniform(-snr.delta_t / 2.0,
+                                   snr.delta_t / 2.0,
+                                   size=self.vsamples)
+        tc = tct + tci * snr.delta_t + float(snr.start_time) - dt
+
+        # Update the current proposed times and the marginalization values
+        self.marginalize_vector_params['tc'] = tc
+        self._current_params['tc'] = tc
+
+        # Update the importance weights for each vector sample
+        logw = self.marginalize_vector_weights - logweight[tci]
+        self.marginalize_vector_weights = logw - logsumexp(logw)
+
+    def draw_sky_times(self, snrs):
+        """ Draw ra, dec, and tc together using SNR timeseries to determine
+        monte-carlo weights.
+        """
+        # First setup
+        # precalculate dense sky grid and make dict and or array of the results
+        if not hasattr(self, 'tinfo'):
+            logging.info('pregenerating sky pointings')
+            size = int(1e6)
+            logging.info('drawing samples')
+            ra = self.marginalized_vector_priors['ra'].rvs(size=size)['ra']
+            dec = self.marginalized_vector_priors['dec'].rvs(size=size)['dec']
+            tcmin, tcmax = self.marginalized_vector_priors['tc'].bounds['tc']
+            tcave = (tcmax + tcmin) / 2.0
+            ifos = list(snrs.keys())
+            d = {ifo: Detector(ifo, reference_time=tcave) for ifo in ifos}
+
+            # What data structure to hold times? Dict of offset -> list?
+            logging.info('sorting into time delay dict')
+            dts = []
+            for i in range(len(ifos) - 1):
+                dt = d[ifos[0]].time_delay_from_detector(d[ifos[i+1]],
+                                                         ra, dec, tcave)
+                dt = numpy.rint(dt / snrs[ifos[0]].delta_t)
+                dts.append(dt)
+
+            dtc = d[ifos[0]].time_delay_from_earth_center(ra, dec, tcave)
+
+            dmap = {}
+            for i, t in enumerate(tqdm.tqdm(zip(*dts))):
+                v = ra[i], dec[i], dtc[i]
+                if t not in dmap:
+                    dmap[t] = []
+                dmap[t].append(v)
+
+            self.tinfo = ifos, dmap, d, tcmin, tcmax
+
+        ifos, dmap, d, tcmin, tcmax = self.tinfo
+
+        # draw times from each snr time series
+        # Is it worth doing this if some detector has low SNR?
+        sref = None
+        iref = None
+        idx = []
+        dx = []
+        mcweight = None
+        for ifo in ifos:
+            snr = snrs[ifo]
+            start = max(tcmin - .031, snr.start_time)
+            end = min(tcmax + .031, snr.end_time)
```

Comment:
Please define a variable (maybe `EARTH_CROSSING_TIME`) for 0.031 and use that here, so we know what this is.